### PR TITLE
Add devise

### PR DESCRIPTION
**Pull request by scottalexandra**
add devise for login
added enum roles for user back in 
added authorization for admin companies and dashboard controllers! Yay!
fixed failing tests!
All tests passing!!!! WOOT WOOT!

### CODE REVIEW

File: spec/controllers/admin/companies_controller_spec.rb [hunk 1]
```diff
@@ -5,6 +5,9 @@
 
   before(:each) do
     @company = create(:company)
+    admin = User.create(email: "admin@example.com", password: "password", role: 1)
+    allow_any_instance_of(ApplicationController).to receive(:current_user)
+      .and_return(admin)
   end
```
Comment (scottalexandra) [inline spec/controllers/admin/companies_controller_spec.rb:11]: We had to stub out the current user as an admin in order to test these controller actions because of the authorization that we added on the admin companies controller. FYI

File: app/models/user.rb [hunk 2]
```diff
@@ -1,5 +1,8 @@
 class User < ActiveRecord::Base
-  has_secure_password
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable
```
Comment (cluhring) [inline app/models/user.rb:5]: Looks like devise gives us a lot o stuff.  Would love to hear more about it.  Also, I did a bit a research into Active Admin for the_pivot and remember devise was a big part of that gem.

File: app/views/devise/confirmations/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend confirmation instructions" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (cluhring) [inline app/views/devise/confirmations/new.html.erb:16]: wow! you two got a lot done this afternoon!

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<p>Welcome <%= @email %>!</p>
+
+<p>You can confirm your account email through the link below:</p>
+
+<p><%= link_to 'Confirm my account', confirmation_url(@resource, confirmation_token: @token) %></p>
```
Comment (cluhring) [inline app/views/devise/mailer/confirmation_instructions.html.erb:5]: sends email?

File: app/views/devise/mailer/unlock_instructions.html.erb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @resource.email %>!</p>
+
+<p>Your account has been locked due to an excessive number of unsuccessful sign in attempts.</p>
+
+<p>Click the link below to unlock your account:</p>
+
+<p><%= link_to 'Unlock my account', unlock_url(@resource, unlock_token: @token) %></p>
```
Comment (cluhring) [inline app/views/devise/mailer/unlock_instructions.html.erb:7]: what! devise is amazaballz!

File: app/views/devise/passwords/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    <%= f.label :password_confirmation, "Confirm new password" %><br />
+    <%= f.password_field :password_confirmation, autocomplete: "off" %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Change my password" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (cluhring) [inline app/views/devise/passwords/edit.html.erb:22]: what if they choose a password that they previously used?

File: app/views/devise/passwords/new.html.erb [hunk 7]
```diff
@@ -0,0 +1,16 @@
+<h2>Forgot your password?</h2>
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :post }) do |f| %>
+  <%= devise_error_messages! %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
```
Comment (cluhring) [inline app/views/devise/passwords/new.html.erb:8]: what does autofucus do?

File: app/views/devise/registrations/edit.html.erb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+<%= form_for(resource, as: resource_name, url: registration_path(resource_name), html: { method: :put }) do |f| %>
+  <%= devise_error_messages! %>
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <% if devise_mapping.confirmable? && resource.pending_reconfirmation? %>
+    <div>Currently waiting confirmation for: <%= resource.unconfirmed_email %></div>
```
Comment (cluhring) [inline app/views/devise/registrations/edit.html.erb:12]: awaiting?

File: app/views/devise/registrations/new.html.erb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password %>
+    <% if @validatable %>
+    <em>(<%= @minimum_password_length %> characters minimum)</em>
```
Comment (cluhring) [inline app/views/devise/registrations/new.html.erb:14]: where is min length defined?

File: app/views/devise/shared/_links.html.erb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != 'unlocks' %>
+  <%= link_to "Didn't receive unlock instructions?", new_unlock_path(resource_name) %><br />
+<% end -%>
+
+<%- if devise_mapping.omniauthable? %>
+  <%- resource_class.omniauth_providers.each do |provider| %>
+    <%= link_to "Sign in with #{provider.to_s.titleize}", omniauth_authorize_path(resource_name, provider) %><br />
+  <% end -%>
+<% end -%>
```
Comment (cluhring) [inline app/views/devise/shared/_links.html.erb:25]: is this standard devise stuff?  omniauth?

File: app/views/devise/unlocks/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true %>
+  </div>
+
+  <div class="actions">
+    <%= f.submit "Resend unlock instructions" %>
+  </div>
+<% end %>
+
+<%= render "devise/shared/links" %>
```
Comment (cluhring) [inline app/views/devise/unlocks/new.html.erb:16]: this page looks similar to app/views/devise/confirmations/new.html.erb.  Partial possible?

File: config/environments/development.rb [hunk 12]
```diff
@@ -18,4 +18,6 @@
     :secret_access_key => ENV['AWS_SECRET_ACCESS_KEY']
   }
 }
+
+  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (cluhring) [inline config/environments/development.rb:22]: cool!

File: config/initializers/devise.rb [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+  # The secret key used by Devise. Devise uses this key to generate
+  # random tokens. Changing this key will render invalid all existing
+  # confirmation, reset password and unlock tokens in the database.
+  # config.secret_key = '372f9dbc7a124a9c0136bf0be40af73206312744ac8741c2101b322cc939bc144266f0c93b7a7365390c900ac5f30b2e6a01d5d80afdb72f895e6d3ef2cd3f85'
+
+  # ==> Mailer Configuration
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
Comment (cluhring) [inline config/initializers/devise.rb:13]: nice cryptic message :)

File: config/initializers/devise.rb [hunk 14]
```diff
@@ -0,0 +1,259 @@
+# Use this hook to configure devise mailer, warden hooks and so forth.
+# Many of these configuration options can be set straight in your model.
+Devise.setup do |config|
+  # The secret key used by Devise. Devise uses this key to generate
+  # random tokens. Changing this key will render invalid all existing
+  # confirmation, reset password and unlock tokens in the database.
+  # config.secret_key = '372f9dbc7a124a9c0136bf0be40af73206312744ac8741c2101b322cc939bc144266f0c93b7a7365390c900ac5f30b2e6a01d5d80afdb72f895e6d3ef2cd3f85'
+
```
Comment (cluhring) [inline config/initializers/devise.rb:8]: is this our secret key?  not very secret :disappointed:

File: config/initializers/devise.rb [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+  #
+  #     mount MyEngine, at: '/my_engine'
+  #
+  # The router that invoked `devise_for`, in the example above, would be:
+  # config.router_name = :my_engine
+  #
+  # When using omniauth, Devise cannot automatically set Omniauth path,
+  # so you need to do it manually. For the users scope, it would be:
+  # config.omniauth_path_prefix = '/my_engine/users/auth'
+end
```
Comment (cluhring) [inline config/initializers/devise.rb:259]: devise is a beast!

File: config/locales/devise.en.yml [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+  errors:
+    messages:
+      already_confirmed: "was already confirmed, please try signing in"
+      confirmation_period_expired: "needs to be confirmed within %{period}, please request a new one"
+      expired: "has expired, please request a new one"
+      not_found: "not found"
+      not_locked: "was not locked"
+      not_saved:
+        one: "1 error prohibited this %{resource} from being saved:"
+        other: "%{count} errors prohibited this %{resource} from being saved:"
```
Comment (cluhring) [inline config/locales/devise.en.yml:60]: reminds me of our printer class from mastermind ;)

File: coverage/.last_run.json [hunk 17]
```diff
@@ -0,0 +1,5 @@
+{
+  "result": {
+    "covered_percent": 95.87
+  }
+}
```
Comment (cluhring) [inline coverage/.last_run.json]: we're better than that. global gitignore?

File: coverage/.resultset.json [hunk 18]
```diff
@@ -0,0 +1,268 @@
...
+        null,
+        1,
+        1,
+        null,
+        null
+      ]
+    },
+    "timestamp": 1430264034
+  }
+}
```
Comment (cluhring) [inline coverage/.resultset.json]: global gitignore?

File: coverage/assets/0.8.0/application.css [hunk 19]
```diff
@@ -0,0 +1,799 @@
...
+  background-color: #efefef; }
+.source_table .never:nth-child(even) {
+  background-color: #f4f4f4; }
+.source_table .skipped:nth-child(odd) {
+  background-color: #fbf0c0; }
+.source_table .skipped:nth-child(even) {
+  background-color: #fbffcf; }
+
+
+
```
Comment (cluhring) [inline coverage/assets/0.8.0/application.css:799]: all devise css?

File: coverage/index.html [hunk 20]
```diff
@@ -0,0 +1,2239 @@
...
+          
+          
+          <code class="ruby">end</code>
+        </li>
+      
+    </ol>
+  </pre>
+</div>
+      
+
```
Comment (cluhring) [inline coverage/index.html]: Orion, did you ever find this page?  Wait, is this from Devise or simplecov?  works ok?  I'd love to get a quick refresher on simplecov, tomorrow?

File: db/migrate/20150428201906_devise_create_users.rb [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+
+      t.timestamps
+    end
+
+    add_index :users, :email,                unique: true
+    add_index :users, :reset_password_token, unique: true
+    # add_index :users, :confirmation_token,   unique: true
+    # add_index :users, :unlock_token,         unique: true
+  end
+end
```
Comment (cluhring) [inline db/migrate/20150428201906_devise_create_users.rb:42]: Was pretty easy to get working out of the box?

File: spec/features/admin_logins_spec.rb [hunk 22]
```diff
@@ -3,47 +3,38 @@
...
+      fill_in 'user[email]', with: user.email
+      fill_in 'user[password]', with: user.password
+      click_link_or_button('Log in')
 
-      expect(current_path).to_not eq(admin_login_path)
-      expect(current_path).to_not eq(admin_dashboard_index_path)
+      expect(current_path).to eq(root_path)
     end
   end
 end
```
Comment (cluhring) [inline spec/features/admin_logins_spec.rb:121]: Did you make any tests for the mailers?  Are those working yet?

File: app/views/devise/mailer/confirmation_instructions.html.erb (hunk 4 above)
Comment (oorion) [inline app/views/devise/mailer/confirmation_instructions.html.erb:5]: I think a lot of this is stuff that devise builds for you.  When we get to the mailer part, we'll look at what it gives us in more detail.

File: app/views/devise/passwords/edit.html.erb (hunk 6 above)
Comment (oorion) [inline app/views/devise/passwords/edit.html.erb:22]: I'm not sure how devise handles this yet.

File: app/views/devise/passwords/new.html.erb (hunk 7 above)
Comment (oorion) [inline app/views/devise/passwords/new.html.erb:8]: I'm not sure offhand how it works.  Devise built this for us.

File: app/views/devise/registrations/edit.html.erb (hunk 8 above)
Comment (oorion) [inline app/views/devise/registrations/edit.html.erb:12]: Not sure.  I feel like we probably won't touch a lot of this unless we need to customize devise.

File: app/views/devise/registrations/new.html.erb (hunk 9 above)
Comment (oorion) [inline app/views/devise/registrations/new.html.erb:14]: Not sure without looking into more details on devise.

File: config/initializers/devise.rb (hunk 14 above)
Comment (oorion) [inline config/initializers/devise.rb:8]: If and when we end up using this, we'll change it :)

File: coverage/.last_run.json (hunk 17 above)
Comment (oorion) [inline coverage/.last_run.json]: I put it in my gitignore but, for some reason, it added it.  I'll remove this.

File: coverage/.resultset.json (hunk 18 above)
Comment (oorion) [inline coverage/.resultset.json]: I have this in my gitignore:
coverage

Is that sufficient?

File: coverage/assets/0.8.0/application.css (hunk 19 above)
Comment (oorion) [inline coverage/assets/0.8.0/application.css:799]: yes.

File: coverage/index.html (hunk 20 above)
Comment (oorion) [inline coverage/index.html]: I think I wasn't able to view this from localhost because that's the development environment.  Instead, you have to type open coverage/index.html from you terminal.

File: db/migrate/20150428201906_devise_create_users.rb (hunk 21 above)
Comment (oorion) [inline db/migrate/20150428201906_devise_create_users.rb:42]: Yeah, this was all generated for us.

File: spec/features/admin_logins_spec.rb (hunk 22 above)
Comment (oorion) [inline spec/features/admin_logins_spec.rb:121]: Haven't worked on the mailers yet.  That'll be another pr.

File: app/views/devise/unlocks/new.html.erb (hunk 11 above)
Comment (oorion) [inline app/views/devise/unlocks/new.html.erb:16]: This is all pregenerated stuff.

File: app/views/devise/shared/_links.html.erb (hunk 10 above)
Comment (oorion) [inline app/views/devise/shared/_links.html.erb:25]: Yes, standard devise code.

File: app/views/devise/confirmations/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,16 @@
+<h2>Resend confirmation instructions</h2>
+
+<%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
```
Comment (bhargavisatpathy) [inline app/views/devise/confirmations/new.html.erb:3]: What is the resource here? Is it from device?

Comment (oorion) [inline app/views/devise/confirmations/new.html.erb:3]: Yes, it's from devise

**Comment by bhargavisatpathy** [comment]
All devise :+1:
